### DISCUSSION TITLE
Git storage sync does not accurately synchronize database

### DISCUSSION THREAD

**Discussion by ghost** [requarks/wiki#6719] [Error / Bug Report]
I have created a git storage for my Wiki.js instance and connected it to a repository on a gitlab server. I then moved an entire folder of wiki pages to a different location within the git repository and synchronized those changes with the gitlab server. Then, Wiki.js succesfully synchronized those changes into its local git repository but the synchronization into Wiki.js's database messes something up because on the webpage I can see that I end up with two copies of the folder that I moved: One in the old location and one in the new location. None of the buttons in the admin settings (Force sync, ...) changes anything about this. 

So, while the synchronization on the git side of things seems to work, the transfer of changes into Wiki.js's database does not work correctly. The database is not in sync with the local git repo.

**Comment by ghost** [comment]
Am I the only person experiencing this? I like to edit files of the wiki locally on my computer and sync them to Gitlab and from there to wikijs. However, again and again, syncing from the local repository on the wikijs machine to the wikijs database messes something up. Files are ignored or renamed. This kind of makes the git sync option useless to me.

**Comment by jackadam1981** [comment]
Try:
1.storage.title  --  purge local repository
2.utilities.title  --  utilities.contentTitle -- rebuild page tree
3.storage.title -- storage.syncDirpull 

Clear local files and resynchronize.

**Comment by Yayg** [comment]
:+1: 
I'm experiencing the same issue.

@FriedrichTuttas described the problem quite well.
To reproduce:

1. Add a git repository as storage backend (with bi-directionnal sync)
2. Create pages in wikijs webUI
3. Move theses pages in git
4. Look at the tree in the "Browse" section

I tried what you said @jackadam1981 but it didn't work.
To me there is a feature missing to drop the database links and rebuild after each sync with the git repo because if your source of truth is git there should not be any untracked changes in the DB (which is the case if you try to "Add Untracked Changes").

Any

**Reply by Yayg** [reply]
FWIW: When you delete a page after going through @jackadam1981's instructions and then try to delete a moved page you get an error saying:
```
fatal: pathspec './mypage.md' did not match any files 
```

**Comment by levino** [comment]
Hey there, might be a bit off topic but I am using wikijs for the first time and I do not yet fully understand the relation between the database and the (local) git repository. I think writing a two way mapping between a git repository and a database state is a very hard (expensive at least in terms of effort) task. Are the technical details explained somewhere in a complete and easy to digest manner? Or should I rather read some code in order to understand what is going on in detail and what the requirements and features of the functionality are? Thanks for any pointers or input.

**Comment by ghost** [comment]
Was anyone able to find an effective way to purge entire directories from the database? We're dealing with a similar issue here where instead of moving the content, we attempted to purge several entire directories from the site. Git did it's thing successfully and removed all the content as desired. But the database only deleted about 10% of what we needed deleted. Now, the directories and pages underneath them are seemingly stuck in the site until we purge manually from the database.

I did discover that if you attempt to delete a page, it will error out, but afterwards, actually remove the page after it realizes it can't find it. However, this is not a viable solution to go this route as we have thousands of pages that need deleted.

**Comment by zmberber** [comment]
bumping this.  i have had problems in my wiki.js instance with this as well.  i would like to test this myself soon again, with the newest version.

to me, this is a rather big issue.  you should be able to edit things in the git and in wiki.js and be able to synchronize bidirectionally, isn't that the idea of the synchronization?  with this issue, it seems like this feature is broken/experimental/buggy, which has been my experience.

**Reply by levino** [reply]
Tbh, I just use Docusaurus/starlight and it is a bliss.

**Reply by zmberber** [reply]
how is this relevant to our problem?  i don't understand.  or are you just saying that you use different software alltogether instead of wiki.js?  documentation software does not fulfill the same purpose as wiki software, i don't see the software you mentioned as viable alternstives.  or are you suggesting a sync setup for wiki.js via this other software that solves the problem?

**Reply by levino** [reply]
I think the idea to get a working two way sync between a database and git working is a dead end. Some people might be interested in my recommendation to just drop the database approach and go with git and markdown only. It serves the same purpose and achieves the same result: Collaborative documentation.

**Reply by zmberber** [reply]
this is true if everyone works with git.  i would have done such an option if everyone was comfortable with git, etc.

however, a lot of people work with a group of people where a web interface for editing is needed.  you want everyone, even non-IT people to be able to use it.

also, wiki.js offers roles on different subdirectories with different read/write permissions etc., which is very handy and not that easy with a pure git setup.  wiki.js also has so many other nice features like SSO as a package, it works very well.

Maybe I am being naive, but I don't see why being able to force sync in one direction shouldn't work (i.e., I do not see it as a dead end).

**Reply by levino** [reply]
What is the point of syncing only in one direction? I would say this feature makes only sense if there is a two way sync. And if this is achieved, one has effectively built an "intuitive git interface" which so many people tried and failed before. This is why I think it is probably impossible. Trying to build it smacks of hybris. The bugginess of the feature supports my assumptions.
